<!doctype html><html lang="zh-tw">
<head>
    <meta http-equiv="refresh" content="0; URL=https://openhome.cc/Gossip/Java/'https://openhome.cc/zh-tw/java/exception/try-catch/'">
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="description" content="在 使用 try、catch 的Average範例中，雖然沒有撰寫try、catch語句，照樣可以編譯執行，Java初學者往往不理解的是，如果如下撰寫，編譯卻會錯誤？
      
      ...">

<meta property="og:locale" content="zh_TW">
<meta property="og:title" content="例外繼承架構">
<meta property="og:type" content="article">
<meta property="og:url" content="https://openhome.cc/Gossip/Java/ThrowableHierarchy.html">
<meta property="og:image" content="https://openhome.cc/Gossip/images/caterpillar_small.jpg">
<meta property="og:site_name" content="OPENHOME.CC">
<meta property="og:description" content="在 使用 try、catch 的Average範例中，雖然沒有撰寫try、catch語句，照樣可以編譯執行，Java初學者往往不理解的是，如果如下撰寫，編譯卻會錯誤？
      
      ...">


    <title>例外繼承架構</title>

<link rel="stylesheet" href="../css/pure-0.6.0/pure-min.css">

    <!--[if lte IE 8]>
        <link rel="stylesheet" href="../css/layouts/side-menu-old-ie.css">
    <![endif]-->
    <!--[if gt IE 8]><!-->
        <link rel="stylesheet" href="../css/layouts/side-menu.css">
    <!--<![endif]-->
  

     <link rel="stylesheet" href="../css/caterpillar.css">
     <script async src="../google-code-prettify/run_prettify.js"></script>
<!-- 網頁層級廣告 --><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><script>(adsbygoogle =window.adsbygoogle || []).push({google_ad_client: "ca-pub-9750319131714390",enable_page_level_ads: true });</script></head>
<body>

<div id="layout">
    <!-- Menu toggle -->
    <a href="ThrowableHierarchy.html#menu" id="menuLink" class="menu-link">
        <!-- Hamburger icon --><span></span>
        
    </a>
    <div id="menu">
        <div class="pure-menu">
            <a class="pure-menu-heading" href="index.html">回 Java 目錄</a>
            <ul class="pure-menu-list">
                <br><div class="social" style="text-align: center;"><a href="http://twitter.com/caterpillar"><img title="Twitter" alt="Twitter" src="../images/twitter.png"></a> <a href="http://www.facebook.com/openhome.cc"><img title="Facebook" alt="Facebook" src="../images/facebook.png"></a> </div><br><div id="search box"><script>(function() {var cx = 'partner-pub-9750319131714390:3926766884';var gcse = document.createElement('script');gcse.type = 'text/javascript';gcse.async = true;gcse.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') + '//www.google.com/cse/cse.js?cx=' + cx;var s = document.getElementsByTagName('script')[0];s.parentNode.insertBefore(gcse, s);})();</script><gcse:searchbox-only></gcse:searchbox-only></div><br><div class="ad" style="text-align: center;"><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
<!-- 2015 新版型 160 x 600 廣告 -->
<ins class="adsbygoogle"
     style="display:inline-block;width:160px;height:600px"
     data-ad-client="ca-pub-9750319131714390"
     data-ad-slot="3747048883"></ins>
<script>
(adsbygoogle = window.adsbygoogle || []).push({});
</script></div></li>
            </ul>
        </div>
    </div>

    <main id="main">
        <header class="header">
            <h1>例外繼承架構</h1>
        </header>

        <article class="content"><br><div class="ad-3" style="text-align: center;"><script async="" src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><!-- 2015 新版型回應式廣告 --><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-9750319131714390" data-ad-slot="7104125683" data-ad-format="auto"></ins><script>(adsbygoogle = window.adsbygoogle || []).push({});</script></div> 
            <code> </code>
      <p>在 <a href="TryCatch.html">使用 <code>try</code>、<code>catch</code></a> 的<code>Average</code>範例中，雖然沒有撰寫<code>try</code>、<code>catch</code>語句，照樣可以編譯執行，Java初學者往往不理解的是，如果如下撰寫，編譯卻會錯誤？</p>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="為什麼一定要處理java.io.IOException？" alt="為什麼一定要處理java.io.IOException？"

        src="images/ThrowableHierarchy-1.PNG" /></div></div><br>
      <br>
      單純就語法上想解決這個錯誤訊息有兩種方式，一是使用<code>try</code>、<code>catch</code>包裹<code>System.in.read()</code>，二是在<code>main()</code>方法旁宣告<code>throws java.io.IOException</code>。簡單來說，API設計者在<code>System.in.read()</code>的實作中，某些條件成立時會引發錯誤，而且認為呼叫<code>System.in.read()</code>方法的客戶端有能力處理錯誤，要求編譯器提醒客戶端必須明確處理錯誤，不然不可通過編譯。例如，若如下撰寫就可以通過編譯：<br>
      <br>
      <div style="margin-left: 40px;"><code>try {<br>
          &nbsp;&nbsp;&nbsp; int ch = System.in.read();<br>
          &nbsp;&nbsp;&nbsp; System.out.println(ch);<br>
        </code> <code>} catch(java.io.IOException ex) {<br>
          &nbsp;&nbsp;&nbsp; ex.printStackTrace() ;<br>
        </code> <code>}</code></div>
      <br>
      <code> Average</code>範例與這邊的例子，程式都有可能發生錯誤，為什麼編譯器單單就只要求這邊的範例，一定要處理錯誤呢？要瞭解這個問題，得先瞭解那些錯誤物件的繼承架構：<br>
      <br>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="Throwable繼承架構圖" alt="Throwable繼承架構圖" src="images/ThrowableHierarchy-2.PNG" /></div></div><br>
      <br>
      首先要瞭解錯誤會被包裝為物件，這些物件都是可拋出的（稍後會介紹<code>throw</code>語法，就會瞭解如何拋出錯誤物件），因此可拋出的物件都繼承自<code>java.lang.Throwable</code>類別，並可使用<code>try</code>、<code>catch</code>進行捕捉，<code>Throwable</code>定義了取得訊息、堆疊追蹤（Stack Trace）等方法，它有兩個子類別：<code>java.lang.Error</code>與<code>java.lang.Exception</code>。<br>
      <br>
      <code>Error</code>與其子類別實例代表嚴重系統錯誤，例如硬體層面錯誤、JVM錯誤或記憶體不足等問題。雖然也可以使用<code>try</code>、<code>catch</code>來處理<code>Error</code>物件，但並不建議，發生嚴重系統錯誤時，Java應用程式本身是無力回復的。舉例來說，若JVM所需記憶體不足，如何撰寫程式要求作業系統給予JVM更多記憶體呢？<code>Error</code>物件拋出時，基本上不用處理，任其傳播至JVM為止，或者是最多留下日誌訊息，供開發者除錯時作為修正程式時的線索。<br>
      <br>
      如果拋出了<code>Throwable</code>物件，而程式中沒有任何<code>catch</code>捕捉到物件，最後由JVM捕捉到的話，那JVM基本處理就是顯示錯誤物件包裹之訊息並中斷程式。<br>
      <br>
      程式設計本身的錯誤，建議使用<code>Exception</code>或其子類別實例來表現，所以通常稱錯誤處理為例外處理（Exception handling）。<br>
      <br>
      單就語法與繼承架構上來說，如果某個方法宣告會拋出<code>Throwable</code>、<code>Exception</code>或子類別實例，但又不屬於<code>Error</code>或<code>java.lang.RuntimeException</code>或其子類別實例，你就必須明確使用<code>try</code>、<code>catch</code>語法加以處理，或者在方法用<code>throws</code>宣告這個方法會拋出例外，否則會編譯失敗。<br>
      <br>
      例如，先前呼叫<code>System.in.read()</code>時，<code>in</code>其實是<code>System</code>的靜態成員，其型態為<code>java.io.InputStream</code>，如果你查詢API文件，可以看到<code>InputStream</code>的<code>read()</code>方法被API設計者宣告為：<br>
      <br>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="read()宣告會拋出IOException" alt="read()宣告會拋出IOException" src="images/ThrowableHierarchy-3.PNG" /></div></div><br>
      <br>
      <code>IOException</code>是<code>Exception</code>的直接子類別，所以編譯器要求你明確使用語法加以處理。<code>Exception</code>或其子物件，但非屬於<code>RuntimeException</code>或其子物件，稱為受檢例外（Checked Exception），受誰檢查？受編譯器檢查！受檢例外存在之目的，在於API設計者實作某方法時，某些條件成立時會引發錯誤，而且認為呼叫方法的客戶端有能力處理錯誤，要求編譯器提醒客戶端必須明確處理錯誤，不然不可通過編譯，API客戶端無權選擇要不要處理。<br>
      <br>
      屬於<code>RuntimeException</code>衍生出來的類別實例，代表API設計者實作某方法時，某些條件成立時會引發錯誤，而且認為API客戶端應該在呼叫方法前做好檢查，以避免引發錯誤，之所以命名為執行時期例外，是因為編譯器不會強迫一定得在語法上加以處理，亦稱為非受檢例外（Unchecked Exception）。<br>
      <br>
      例如使用陣列時，若存取超出索引就會拋出<code>ArrayIndexOutOfBoundsException</code>，但編譯器並沒有強迫你在語法上加以處理，這是因為<code>ArrayIndexOutOfBoundsException</code>是一種<code>RuntimeException</code>，可以在API文件的開頭找到繼承架構圖：<br>
      <br>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="ArrayIndexOutOfBoundsException是一種RuntimeException" alt="ArrayIndexOutOfBoundsException是一種RuntimeException"

        src="images/ThrowableHierarchy-4.PNG" /></div></div><br>
      <br>
      在 <a href="TryCatch.html">使用 <code>try</code>、<code>catch</code></a> 的<code>Average</code>範例中，因為<code>InputMismatchException</code>設計為一種<code>RuntimeException</code>：<br>
      <br>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="InputMismatchException是一種RuntimeException" alt="InputMismatchException是一種RuntimeException"

        src="images/ThrowableHierarchy-5.PNG" /></div></div><br>
      <br>
      因此即使實際上還是有可能引發例外，但編譯器並不強迫你得處理例外，你有權利選擇是否使用<code>try</code>、<code>catch</code>處理，Java對於<code>RuntimeException</code>的態度是，這是一種臭蟲，也就是呼叫方法前沒有做好前置檢查，應該予以修正，使得呼叫方法時不會引發錯誤，如果真要以<code>try</code>、<code>catch</code>處理，建議是日誌或呈現友善訊息，像是 <a

        href="TryCatch.html">使用 <code>try</code>、<code>catch</code></a> 的<code>Average2</code>的作法就是個例子。<br>
      <br>
      雖然有些小題大作，不過 <a href="TryCatch.html">使用 <code>try</code>、<code>catch</code></a><code> </code>的<code>Average3</code>若要避免出現<code>InputMismatchException</code>，應該是取得使用者的字串輸入之後，檢查是否為數字格式，若是再轉換為<code>int</code>整數，若否則提醒使用者做正確格式輸入，例如：<br>
      <br>
      <pre class="prettyprint"><code lang="java">package cc.openhome;

import java.util.Scanner;

public class Average4 {

    public static void main(String[] args) {
        double sum = 0;
        int count = 0;
        int number;
        while(true) {
            number = nextInt();
            if(number == 0) {
                break;
            }
            sum += number;
            count++;
        }
        System.out.printf("平均 %.2f%n", sum / count);
    }

    static Scanner scanner = new Scanner(System.in);

    static int nextInt() {
        String input = scanner.next();
        while(!input.matches("\\d*")) {
            System.out.println("請輸入數字");
            input = scanner.next();
        }
        return Integer.parseInt(input);
    }
}
</code></pre> <br>
      除了瞭解<code>Error</code>與<code>Exception</code>的區別，以及<code>Exception</code>、<code>RuntimeException</code>的分別之外，使用<code>try</code>、<code>catch</code>捕捉例外物件時也要注意，如果父類別例外物件在子類別例外物件前被捕捉，則<code>catch</code>子類別例外物件的區塊將永遠不會被執行，編譯器會檢查出這個錯誤。例如：<br>
      <br>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="瞭解例外繼承架構是必要的" alt="瞭解例外繼承架構是必要的" src="images/ThrowableHierarchy-6.PNG" /></div></div><br>
      <br>
      要完成這個程式的編譯，必須更改例外物件捕捉的順序。例如：<br>
      <br>
      <div style="margin-left: 40px;"><code>try {</code><br>
        <code>&nbsp;&nbsp;&nbsp; System.in.read();</code><br>
        <code>} catch(java.io.IOException e) {</code><br>
        <code>&nbsp;&nbsp;&nbsp; e.printStackTrace();</code><br>
        <code>} catch(Exception e) {</code><br>
        <code>&nbsp;&nbsp;&nbsp; e.printStackTrace();</code><br>
        <code>}</code></div>
      <br>
      經常地，你會發現到數個型態的<code>catch</code>區塊在作相同的事情，這種情況常發生在某些例外都要進行日誌記錄的情況。例如：<br>
      <br>
      <div style="margin-left: 40px;"><code>try {</code><br>
        <code>&nbsp;&nbsp;&nbsp; 作一些事...</code><br>
        <code>} catch(IOException e) {</code><br>
        <code>&nbsp;&nbsp;&nbsp; e.printStackTrace();</code><br>
        <code>} catch(InterruptedException e) {</code><br>
        <code>&nbsp;&nbsp;&nbsp; e.printStackTrace();</code><br>
        <code>} catch(ClassCastException e) {</code><br>
        <code>&nbsp;&nbsp;&nbsp; e.printStackTrace();</code><br>
        <code>}</code></div>
      &nbsp;&nbsp; <br>
      然而<code>catch</code>例外後的區塊內容重複了，撰寫時不僅無趣且對維護並沒有幫助。在JDK7開始，可以如下使用多重捕捉（multi-cath）語法：<br>
      <br>
      <div style="margin-left: 40px;"><code>try {</code><br>
        <code>&nbsp;&nbsp;&nbsp; 作一些事...</code><br>
        <code>} catch(IOException | InterruptedException | ClassCastException e) {</code><br>
        <code>&nbsp;&nbsp;&nbsp; e.printStackTrace();</code><br>
        <code>}</code></div>
      &nbsp;&nbsp; <br>
      這樣的撰寫方式簡潔許多，<code>catch</code>區塊會在發生<code>IOException</code>或<code>InterruptedException</code>或<code>ClassCastException</code>時執行，不過仍得注意例外的繼承，<code>catch</code>括號中列出的例外不得有繼承關係，否則會發生編譯錯誤：<br>
      <br>
      <div class="pure-g"><div class="pure-u-1"><img class="pure-img-responsive" title="多重捕捉時也得注意例外繼承架構" alt="多重捕捉時也得注意例外繼承架構" src="images/ThrowableHierarchy-7.PNG" /></div></div><br>
      <br>
      <br>
      <p> </p>
      <br>
      <ul>
      </ul><br><br><div class="ad336-280" style="text-align: center;"><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><!-- 2015 新版型廣告 336 x 280 --><ins class="adsbygoogle" style="display:inline-block;width:336px;height:280px" data-ad-client="ca-pub-9750319131714390" data-ad-slot="9976409681"></ins><script>(adsbygoogle = window.adsbygoogle || []).push({});</script></div><br><div class="recommend" style="text-align: center;"><hr><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script><!-- 自動大小回應相符內容 --><ins class="adsbygoogle" style="display:block" data-ad-client="ca-pub-9750319131714390" data-ad-slot="4953478487" data-ad-format="autorelaxed"></ins><script>(adsbygoogle = window.adsbygoogle || []).push({});</script></div></body>
<script src="../js/ui.js"></script>
</html>
<div class="analytics"><script async src="https://www.googletagmanager.com/gtag/js?id=G-QVQQYFSC8J"></script><script>window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}gtag('js', new Date());gtag('config', 'G-QVQQYFSC8J');</script></div>
